extensions/replication/utils/getRedisClient.js

```diff
@@ -1,8 +1,18 @@
 const { RedisClient } = require('arsenal').metrics;
 const config = require('../../../conf/Config');
+const werelogs = require('werelogs');
 
-function getRedisClient(log) {
-    return new RedisClient(config.redis, log);
+const log = new werelogs.Logger('Backbeat:RedisClient');
+werelogs.configure({
+    level: config.log.logLevel,
+    dump: config.log.dumpLevel,
+});
+
+function getRedisClient() {
+    const redisConfig = Object.assign({}, config.redis, {
+        enableOfflineQueue: false,
```

Comment:
The Redis hash key operation `hset` is in the critical path of the replication status processor. I've set `enableOfflineQueue` to `false` here in the case that the connection to Redis is faulty—I don't think it's safe to block the processor while waiting for the Redis connection to reestablish. Perhaps we need to implement some kind of "replay" feature for any missed Redis operations in Backbeat.

Reply:
I'm seeing on the docs that if `enableOfflineQueue` is set to false, when a redis command is attempted to be executed and the connection isn't ready, an error is thrown, whereas if true, it seems like its queued somehow and when "ready", it will execute the commands in the queue.
https://github.com/luin/ioredis/blob/cc7d21ab60dd5b01eb5f8af25da968d89629332c/API.md

So maybe this is something we want enabled and might actually handle the issue of having to "retry". I'm wondering also how long queued commands are kept for

Reply:
If we rely on the offline queue, as I understand it, it will stay in the local queue but we can consider it as it was written to redis, so the replication status processor should not be blocked in this case, considering that the update is pending? Of course it may not go through if the process crashes or anything.

Another idea may be to use redis sentinel, I'm not familiar at all with this but it sounds like it would bring more availability and durability, so it could be considered reliable enough for the purpose?

Just random thoughs.

Reply:
Yes, with the offline queue enabled, it will stay in the local queue. However, in this case, the callback from the Redis command (i.e. [this one](https://github.com/scality/backbeat/pull/231/files#diff-9325b23888750621a938fe9d5714c29aR154)) is not invoked until the connection reestablishes, thereby blocking the processor from doing its work during that time. (I've seen it wait for > 30 min, though I'm not sure what the eventual timeout is.)

Reply:
@rahulreddy and I had a discussion concerning this issue. We are thinking a solution would be to push CRR metrics to a dedicated topic, and then have a consumer managing the Redis calls. That way we can avoid a hang when Redis is disconnected, while keeping the offline Redis queue enabled so as to avoid metrics loss. I've created a JIRA ticket for this and can follow-up with a PR to address this unforeseen design issue.

Reply:
sounds good to me